File: scripts/firesmoke.py
#!/usr/bin/env python3
"""
FireSmoke Canada smoke-forecast fetcher for Birchdale.

FireSmoke.ca (UBC Weather Forecast Research Team, BlueSky/HYSPLIT model) has no
JSON API. It publishes each forecast run as a single ~84 MB NetCDF-3 file:

    https://firesmoke.ca/forecasts/current/dispersion.nc

That file is an IOAPI regular lat-lon grid (GDTYP=1) of near-surface PM2.5 in
ug/m^3, dims (TSTEP, LAY, ROW, COL) -- typically 51 hourly steps on a 0.1-deg
grid covering most of North America.

Rather than download all 84 MB, this module parses the ~11 KB header, computes
the grid cell for a given lat/lon, and uses HTTP range requests to read only the
few bytes for that one cell across every timestep. A full pull is ~100 tiny
reads over a keep-alive session (~4 s) and needs nothing beyond `requests` -- no
netCDF4/scipy dependency.

CLI:
    python firesmoke.py                     # Birchdale, pretty table
    python firesmoke.py --lat 49.9 --lon -119.5
    python firesmoke.py --json public/firesmoke_data.json
"""
import argparse
import json
import struct
from datetime import datetime, timedelta, timezone

import requests

from utils import LAT, LON, PACIFIC

FIRESMOKE_URL = "https://firesmoke.ca/forecasts/current/dispersion.nc"

# US EPA PM2.5 -> AQI breakpoints (2024 revision), (C_low, C_high, AQI_low, AQI_high).
# Concentration is truncated to 0.1 ug/m^3 before interpolation, per EPA method.
_AQI_BREAKPOINTS = [
    (0.0, 9.0, 0, 50, "Good"),
    (9.1, 35.4, 51, 100, "Moderate"),
    (35.5, 55.4, 101, 150, "Unhealthy for Sensitive Groups"),
    (55.5, 125.4, 151, 200, "Unhealthy"),
    (125.5, 225.4, 201, 300, "Very Unhealthy"),
    (225.5, 325.4, 301, 500, "Hazardous"),
]


def pm25_to_aqi(conc):
    """Convert a PM2.5 concentration (ug/m^3) to US EPA AQI and category."""
    if conc is None or conc < 0:
        return None, None
    c = int(conc * 10) / 10.0  # truncate to 0.1
    for c_lo, c_hi, a_lo, a_hi, cat in _AQI_BREAKPOINTS:
        if c <= c_hi:
            aqi = round((a_hi - a_lo) / (c_hi - c_lo) * (c - c_lo) + a_lo)
            return aqi, cat
    return 500, "Hazardous"


# ---------------------------------------------------------------------------
# Minimal NetCDF-3 "classic" (CDF\x01) header reader
# ---------------------------------------------------------------------------
_TYPE_SIZE = {1: 1, 2: 1, 3: 2, 4: 4, 5: 4, 6: 8}  # byte,char,short,int,float,double


class _Reader:
    def __init__(self, buf):
        self.b = buf
        self.p = 0

    def u4(self):
        v = struct.unpack_from(">I", self.b, self.p)[0]
        self.p += 4
        return v

    def name(self):
        n = self.u4()
        s = self.b[self.p:self.p + n].decode("latin1")
        self.p += n + ((4 - (n % 4)) % 4)  # skip 4-byte padding
        return s

    def attrs(self):
        self.u4()  # tag (NC_ATTRIBUTE or absent)
        out = {}
        for _ in range(self.u4()):
            an = self.name()
            t = self.u4()
            nv = self.u4()
            sz = _TYPE_SIZE[t] * nv
            raw = self.b[self.p:self.p + sz]
            self.p += sz + ((4 - (sz % 4)) % 4)
            if t == 2:
                out[an] = raw.decode("latin1")
            elif t in (5, 6):
                out[an] = struct.unpack(">%d%s" % (nv, "f" if t == 5 else "d"), raw)
            else:
                out[an] = struct.unpack(">%d%s" % (nv, {1: "b", 3: "h", 4: "i"}[t]), raw)
        return out


def parse_header(hdr):
    """Parse a NetCDF-3 classic header into dims/global-attrs/var-layout."""
    r = _Reader(hdr)
    if r.b[:4] != b"CDF\x01":
        raise ValueError("Not a NetCDF-3 classic file (bad magic %r)" % r.b[:4])
    r.p = 4
    numrecs = r.u4()
    r.u4()  # NC_DIMENSION tag
    dims = [(r.name(), r.u4()) for _ in range(r.u4())]
    gattrs = r.attrs()
    r.u4()  # NC_VARIABLE tag
    variables = {}
    for _ in range(r.u4()):
        vn = r.name()
        dimids = [r.u4() for _ in range(r.u4())]
        r.attrs()  # per-variable attributes (unused here)
        vtype = r.u4()
        vsize = r.u4()
        begin = r.u4()
        variables[vn] = {"dimids": dimids, "type": vtype, "vsize": vsize, "begin": begin}
    return {"numrecs": numrecs, "dims": dims, "gattrs": gattrs, "vars": variables}


def _ioapi_to_utc(yyyyddd, hhmmss):
    """IOAPI date (YYYYDDD) + time (HHMMSS) -> aware UTC datetime."""
    year, doy = divmod(yyyyddd, 1000)
    h, rem = divmod(hhmmss, 10000)
    m, s = divmod(rem, 100)
    base = datetime(year, 1, 1, tzinfo=timezone.utc) + timedelta(days=doy - 1)
    return base + timedelta(hours=h, minutes=m, seconds=s)


# ---------------------------------------------------------------------------
# Forecast extraction
# ---------------------------------------------------------------------------
def fetch_pm25_forecast(lat=LAT, lon=LON, url=FIRESMOKE_URL, timeout=30):
    """
    Return the hourly surface-PM2.5 forecast for the grid cell at (lat, lon).

    Reads only the bytes for that single cell via HTTP range requests.
    Returns a dict: metadata + `series` list of hourly points.
    """
    sess = requests.Session()

    def ranged(a, b):
        resp = sess.get(url, headers={"Range": "bytes=%d-%d" % (a, b)}, timeout=timeout)
        resp.raise_for_status()
        if resp.status_code != 206:
            raise RuntimeError("Server ignored Range request (status %d)" % resp.status_code)
        return resp.content

    meta = parse_header(ranged(0, 16383))
    g = meta["gattrs"]
    ncols, nrows = g["NCOLS"][0], g["NROWS"][0]
    xorig, yorig = g["XORIG"][0], g["YORIG"][0]
    xcell, ycell = g["XCELL"][0], g["YCELL"][0]
    numrecs = meta["numrecs"]

    # Grid cell whose center is nearest (lat, lon). XORIG/YORIG is the SW corner
    # of cell (0,0); cell center = ORIG + (index + 0.5) * CELL.
    col = round((lon - xorig) / xcell - 0.5)
    row = round((lat - yorig) / ycell - 0.5)
    if not (0 <= col < ncols and 0 <= row < nrows):
        raise ValueError(
            "(%.4f, %.4f) is outside the forecast domain "
            "[lon %.1f..%.1f, lat %.1f..%.1f]"
            % (lat, lon, xorig, xorig + ncols * xcell, yorig, yorig + nrows * ycell)
        )

    pm25 = meta["vars"]["PM25"]
    tflag = meta["vars"]["TFLAG"]
    # All record variables share one interleaved record; recsize spans them all.
    recsize = sum(v["vsize"] for v in meta["vars"].values() if v["dimids"] and v["dimids"][0] == 0)
    cell_off = (row * ncols + col) * 4  # LAY=0; float32 element within a record

    series = []
    for t in range(numrecs):
        d, hms = struct.unpack(">2i", ranged(tflag["begin"] + t * recsize,
                                             tflag["begin"] + t * recsize + 7))
        conc = struct.unpack(">f", ranged(pm25["begin"] + t * recsize + cell_off,
                                          pm25["begin"] + t * recsize + cell_off + 3))[0]
        conc = round(max(0.0, conc), 2)
        aqi, cat = pm25_to_aqi(conc)
        utc = _ioapi_to_utc(d, hms)
        series.append({
            "time_utc": utc.strftime("%Y-%m-%dT%H:%M:%SZ"),
            "time_local": utc.astimezone(PACIFIC).strftime("%Y-%m-%d %H:%M %Z"),
            "pm25": conc,
            "aqi": aqi,
            "category": cat,
        })

    peak = max(series, key=lambda p: p["pm25"]) if series else None
    return {
        "source": "FireSmoke.ca (BlueSky/HYSPLIT surface PM2.5)",
        "url": url,
        "lat": lat,
        "lon": lon,
        "cell": {
            "row": row,
            "col": col,
            "center_lat": round(yorig + (row + 0.5) * ycell, 4),
            "center_lon": round(xorig + (col + 0.5) * xcell, 4),
            "resolution_deg": round(xcell, 4),
        },
        "run_start_utc": _ioapi_to_utc(g["SDATE"][0], g["STIME"][0]).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "hours": len(series),
        "peak": peak,
        "series": series,
    }


def main():
    ap = argparse.ArgumentParser(description="Fetch FireSmoke.ca PM2.5 forecast for a point.")
    ap.add_argument("--lat", type=float, default=LAT)
    ap.add_argument("--lon", type=float, default=LON)
    ap.add_argument("--url", default=FIRESMOKE_URL)
    ap.add_argument("--json", metavar="PATH", help="write full forecast to this JSON file")
    args = ap.parse_args()

    fc = fetch_pm25_forecast(args.lat, args.lon, args.url)

    print("FireSmoke PM2.5 forecast for (%.4f, %.4f)" % (fc["lat"], fc["lon"]))
    print("  grid cell center: (%.3f, %.3f)  run start: %s UTC"
          % (fc["cell"]["center_lat"], fc["cell"]["center_lon"], fc["run_start_utc"]))
    pk = fc["peak"]
    if pk:
        print("  peak: %.1f ug/m^3 (AQI %s, %s) at %s"
              % (pk["pm25"], pk["aqi"], pk["category"], pk["time_local"]))
    print()
    print("  %-22s %10s %6s  %s" % ("local time", "PM2.5", "AQI", "category"))
    for p in fc["series"]:
        print("  %-22s %8.1f  %5s  %s" % (p["time_local"], p["pm25"], p["aqi"], p["category"]))

    if args.json:
        with open(args.json, "w") as f:
            json.dump(fc, f, indent=2)
        print("\nWrote %d hours to %s" % (fc["hours"], args.json))


if __name__ == "__main__":
    main()

File: netlify/functions/firesmoke.js
// Netlify Serverless Function: firesmoke.js
// Builds the FireSmoke.ca surface-PM2.5 forecast for a point ON DEMAND (when
// the page is opened) instead of via a daily cron. FireSmoke has no API -- it
// publishes an ~84 MB NetCDF-3 file per run. We parse the ~16 KB header and use
// a single multipart HTTP range request to read just the bytes for one grid
// cell across all timesteps, so the function stays fast and light.
// Mirrors scripts/firesmoke.py (used by the daily email).

const NC_URL = 'https://firesmoke.ca/forecasts/current/dispersion.nc';
const DEFAULT_LAT = 50.038417;
const DEFAULT_LON = -116.892033;

// US EPA PM2.5 -> AQI breakpoints (2024): [C_low, C_high, AQI_low, AQI_high, category]
const AQI_BP = [
  [0.0, 9.0, 0, 50, 'Good'],
  [9.1, 35.4, 51, 100, 'Moderate'],
  [35.5, 55.4, 101, 150, 'Unhealthy for Sensitive Groups'],
  [55.5, 125.4, 151, 200, 'Unhealthy'],
  [125.5, 225.4, 201, 300, 'Very Unhealthy'],
  [225.5, 325.4, 301, 500, 'Hazardous'],
];
function pm25ToAqi(conc) {
  if (conc == null || conc < 0) return [null, null];
  const c = Math.trunc(conc * 10) / 10;
  for (const [clo, chi, alo, ahi, cat] of AQI_BP) {
    if (c <= chi) return [Math.round((ahi - alo) / (chi - clo) * (c - clo) + alo), cat];
  }
  return [500, 'Hazardous'];
}

const TSIZE = { 1: 1, 2: 1, 3: 2, 4: 4, 5: 4, 6: 8 };
function parseHeader(u8) {
  const dv = new DataView(u8.buffer, u8.byteOffset, u8.byteLength);
  let p = 0;
  const u4 = () => { const v = dv.getUint32(p, false); p += 4; return v; };
  const name = () => {
    const n = u4();
    let s = '';
    for (let i = 0; i < n; i++) s += String.fromCharCode(u8[p + i]);
    p += n + ((4 - (n % 4)) % 4);
    return s;
  };
  const attrs = () => {
    u4(); const ne = u4(); const out = {};
    for (let k = 0; k < ne; k++) {
      const an = name(); const t = u4(); const nv = u4(); const sz = TSIZE[t] * nv;
      const vals = [];
      for (let j = 0; j < nv; j++) {
        if (t === 4) vals.push(dv.getInt32(p + j * 4, false));
        else if (t === 5) vals.push(dv.getFloat32(p + j * 4, false));
        else if (t === 6) vals.push(dv.getFloat64(p + j * 8, false));
        else if (t === 3) vals.push(dv.getInt16(p + j * 2, false));
        else vals.push(u8[p + j]);
      }
      p += sz + ((4 - (sz % 4)) % 4);
      out[an] = (t === 2) ? String.fromCharCode(...vals) : vals;
    }
    return out;
  };
  if (String.fromCharCode(u8[0], u8[1], u8[2]) !== 'CDF') throw new Error('not a NetCDF-3 classic file');
  p = 4;
  const numrecs = u4();
  u4(); const nd = u4(); const dims = [];
  for (let i = 0; i < nd; i++) dims.push([name(), u4()]);
  const gattrs = attrs();
  u4(); const nv = u4(); const vars = {};
  for (let i = 0; i < nv; i++) {
    const vn = name(); const ndv = u4(); const dimids = [];
    for (let j = 0; j < ndv; j++) dimids.push(u4());
    attrs(); const type = u4(); const vsize = u4(); const begin = u4();
    vars[vn] = { dimids, type, vsize, begin };
  }
  return { numrecs, dims, gattrs, vars };
}

async function ranged(a, b) {
  const r = await fetch(NC_URL, { headers: { Range: `bytes=${a}-${b}` } });
  if (r.status !== 206) throw new Error(`range read expected 206, got ${r.status}`);
  return new Uint8Array(await r.arrayBuffer());
}

function indexOfSeq(hay, needle, from) {
  outer: for (let i = from; i <= hay.length - needle.length; i++) {
    for (let j = 0; j < needle.length; j++) if (hay[i + j] !== needle[j]) continue outer;
    return i;
  }
  return -1;
}

// Read one float32 from each byte offset, via a single multipart range request.
// Falls back to chunked parallel single-range reads if the server doesn't
// return multipart/byteranges.
async function readFloatsAtOffsets(offsets) {
  const enc = new TextEncoder();
  const rangeHeader = offsets.map(o => `${o}-${o + 3}`).join(',');
  const resp = await fetch(NC_URL, { headers: { Range: `bytes=${rangeHeader}` } });
  const ct = resp.headers.get('content-type') || '';
  if (resp.status === 206 && /multipart\/byteranges/i.test(ct)) {
    const body = new Uint8Array(await resp.arrayBuffer());
    const boundary = enc.encode('--' + ct.match(/boundary=(.+)$/i)[1].trim());
    const crlf2 = enc.encode('\r\n\r\n');
    const byOff = new Map();
    let pos = indexOfSeq(body, boundary, 0);
    while (pos !== -1) {
      const hdrEnd = indexOfSeq(body, crlf2, pos + boundary.length);
      if (hdrEnd === -1) break;
      let hdr = '';
      for (let i = pos + boundary.length; i < hdrEnd; i++) hdr += String.fromCharCode(body[i]);
      const m = hdr.match(/Content-Range:\s*bytes\s+(\d+)-(\d+)\//i);
      const bodyStart = hdrEnd + 4;
      if (m) {
        const start = parseInt(m[1], 10);
        const dv = new DataView(body.buffer, body.byteOffset + bodyStart, 4);
        byOff.set(start, dv.getFloat32(0, false));
      }
      pos = indexOfSeq(body, boundary, bodyStart);
    }
    return offsets.map(o => byOff.get(o));
  }
  // Fallback: parallel single-range reads in chunks of 12.
  const out = new Array(offsets.length);
  for (let i = 0; i < offsets.length; i += 12) {
    const chunk = offsets.slice(i, i + 12);
    const bufs = await Promise.all(chunk.map(o => ranged(o, o + 3)));
    bufs.forEach((b, k) => { out[i + k] = new DataView(b.buffer, b.byteOffset, 4).getFloat32(0, false); });
  }
  return out;
}

function fmtLocal(date) {
  const parts = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'America/Los_Angeles', year: 'numeric', month: '2-digit', day: '2-digit',
    hour: '2-digit', minute: '2-digit', hourCycle: 'h23', timeZoneName: 'short',
  }).formatToParts(date);
  const p = Object.fromEntries(parts.map(x => [x.type, x.value]));
  return `${p.year}-${p.month}-${p.day} ${p.hour}:${p.minute} ${p.timeZoneName}`;
}

function ioapiToUtc(yyyyddd, hhmmss) {
  const year = Math.floor(yyyyddd / 1000), doy = yyyyddd % 1000;
  const h = Math.floor(hhmmss / 10000), m = Math.floor((hhmmss % 10000) / 100), s = hhmmss % 100;
  return new Date(Date.UTC(year, 0, 1) + (doy - 1) * 86400000 + ((h * 3600 + m * 60 + s) * 1000));
}

async function buildForecast(lat, lon) {
  const meta = parseHeader(await ranged(0, 16383));
  const g = meta.gattrs;
  const NCOLS = g.NCOLS[0], NROWS = g.NROWS[0];
  const XORIG = g.XORIG[0], YORIG = g.YORIG[0], XCELL = g.XCELL[0], YCELL = g.YCELL[0];
  const col = Math.round((lon - XORIG) / XCELL - 0.5);
  const row = Math.round((lat - YORIG) / YCELL - 0.5);
  if (col < 0 || col >= NCOLS || row < 0 || row >= NROWS) {
    const err = new Error(`(${lat}, ${lon}) is outside the forecast domain`);
    err.statusCode = 422;
    throw err;
  }
  const pm = meta.vars.PM25;
  const recsize = Object.values(meta.vars).filter(v => v.dimids[0] === 0).reduce((a, v) => a + v.vsize, 0);
  const cellOff = (row * NCOLS + col) * 4;
  const offsets = [];
  for (let t = 0; t < meta.numrecs; t++) offsets.push(pm.begin + t * recsize + cellOff);
  const floats = await readFloatsAtOffsets(offsets);

  const runStart = ioapiToUtc(g.SDATE[0], g.STIME[0]);
  const tstepHours = Math.floor(g.TSTEP[0] / 10000) || 1;
  const series = floats.map((v, t) => {
    const conc = Math.round(Math.max(0, v) * 100) / 100;
    const [aqi, category] = pm25ToAqi(conc);
    const utc = new Date(runStart.getTime() + t * tstepHours * 3600000);
    return {
      time_utc: utc.toISOString().replace('.000Z', 'Z'),
      time_local: fmtLocal(utc),
      pm25: conc, aqi, category,
    };
  });
  const peak = series.reduce((a, b) => (b.pm25 > a.pm25 ? b : a), series[0]);
  return {
    source: 'FireSmoke.ca (BlueSky/HYSPLIT surface PM2.5)',
    url: 'https://firesmoke.ca/forecasts/current/',
    lat, lon,
    cell: {
      row, col,
      center_lat: Math.round((YORIG + (row + 0.5) * YCELL) * 1e4) / 1e4,
      center_lon: Math.round((XORIG + (col + 0.5) * XCELL) * 1e4) / 1e4,
      resolution_deg: Math.round(XCELL * 1e4) / 1e4,
    },
    run_start_utc: runStart.toISOString().replace('.000Z', 'Z'),
    hours: series.length,
    peak, series,
  };
}

exports.handler = async function (event) {
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Content-Type': 'application/json',
    // Cache per edge for 30 min: the model publishes a few times daily.
    'Cache-Control': 'public, max-age=1800',
  };
  if (event.httpMethod === 'OPTIONS') return { statusCode: 200, headers, body: '' };
  if (event.httpMethod !== 'GET') return { statusCode: 405, headers, body: JSON.stringify({ error: 'Method not allowed' }) };
  try {
    const q = event.queryStringParameters || {};
    const lat = q.lat ? parseFloat(q.lat) : DEFAULT_LAT;
    const lon = q.lon ? parseFloat(q.lon) : DEFAULT_LON;
    const fc = await buildForecast(lat, lon);
    return { statusCode: 200, headers, body: JSON.stringify(fc) };
  } catch (e) {
    return {
      statusCode: e.statusCode || 502,
      headers,
      body: JSON.stringify({ error: String(e.message || e) }),
    };
  }
};
